### ISSUE TITLE
Is there a reason to support out of order `axes` for layerNormalization?

### ISSUE THREAD

**Issue by philloooo** [webmachinelearning/webnn#748]
Hi!
I found that on CoreML, when the `axes` are out of order, sometimes the result is wrong. I've also tested with tensorflow, which also returns wrong result.

Is there a reason to support unordered axes? 

We could emulate this by transposing, but if there is no valid use case to support unordered axes, it's simpler to just reject that.

**Comment by lisa0314** [comment]
For ORT: Attribute `axis` defines the first normalization dimension. If rank(X) is r, axis allowed range is [-r, r). Negative value means counting dimensions from the back. So ORT only support last consecutive axes.

**Comment by fdwr** [comment]
To clarify, by "out of order", do you mean *nonincreasing* or *noncontiguous*? I'm guessing you mean *nonincreasing*, and that *increasing noncontiguous* works fine:

- `[2,3]` - increasing and contiguous
- `[1,3]` - increasing and noncontiguous
- `[3,2]` - decreasing and contiguous
- `[3,1]` - decreasing and noncontiguous

~~Could it be a CoreML bug, if it accepts the axes (no error) but silently returns an unexpected result 🐞👀? Or does CoreML sort the axes first before mapping them to the respective other tensors?~~ *Never mind, as CoreML actually takes a `normalizedShape` parameter rather than `axes`*. Logically it would make most sense to me that the axes correspond to the respective dimensions in the `scale` and `bias` tensors, like how the `axes` in `resample` correspond to the respective indices within the `sizes` and `scales` attributes.

I suppose the biggest reason is simply API consistency. Given that `reduce*`, `reverse`, `resample` all support axes that are bidirectional & noncontiguous axes, consistency is elegant.

**Comment by anssiko** [comment]
Sharing the current status with folks who may have missed the [previous meeting discussion](https://www.w3.org/2025/06/05-webmachinelearning-minutes.html#7a84): we agreed to create a repro case to be shared in this issue and with the Core ML development team.

**Comment by fdwr** [comment]
## Interesting cases

I see 3 interesting cases to consider here for axes:

- Noncontiguous (e.g. `1,3`, where 2 is skipped)
- Decreasing order (e.g. `3,2,1`, with decreasing order)
- Non-right-aligned (e.g `0,1,2`, where 3 is excluded for 4D input)

(`reduce*`, `resample`, `reverse` support all the above traits)

Some test cases given 4D input:

Axes | Right-aligned | Contiguous | Increasing order
------|----------------|--------------|-------------------
1,2,3 | 🟢 | 🟢 | 🟢
3,2,1 | 🟢 | 🟢 | .
0,1,2 | . | 🟢 | 🟢
2,1,0 | . | 🟢 | .
1,3 | 🟢 | . | 🟢
1,2 | . | 🟢 | 🟢
2,0 | . | . | .

## API support

Checking a few API's...

- [TF Keras layer normalization](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LayerNormalization) accepts arbitrary axes. So one could normalize along just one axis (e.g. `axis = 2`) or multiple axes (e.g. `axis=[1,2,3]` or `axis=[2,3]` for NHWC layout). So it doesn't need to be contiguous or right-aligned (and possible not even increasing, but I haven't tried out-of-order axes to see how it behaves 🤔). Example snippet:

```python
layer = keras.layers.LayerNormalization(axis=[1, 2, 3])
```

- [CoreML's `LayerNormalizationLayerParams`](https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#layernormalizationlayerparams) takes a `normalizedShape` rather than `axes` (and so [my question](https://www.w3.org/2025/06/05-webmachinelearning-minutes.html#7a84) about whether CoreML sorts the axes is moot).  So axes evidently must be right-aligned and increasing (short of a transpose), but I'm unsure about a contiguity requirement - does the `normalizedShape` just need to be broadcast compatible? e.g. Can you pass input shape = [2,3,4] with normalizedShape = [2,1,4]?

- [ONNX's LayerNorm](https://onnx.ai/onnx/operators/onnx__LayerNormalization.html) only takes a "first axis" range, where axes = [firstAxis ... inputRank-1]. So `axes` must be right-aligned, contiguous, and increasing (short of a transpose).

## Thoughts

To support all the interesting cases (including the TF Keras examples including single axis = 2 and non-contiguous cases), you'd still need a transpose anyway to implement it in CoreML, right?
